finality-aleph/src/testing/data_io.rs

```diff
@@ -0,0 +1,385 @@
+
+    let (block_requester, block_requests_rx, _justification_requests_rx) =
+        TestBlockRequester::new();
+    let (test_network, _network_rx, network_tx) = TestNetwork::new();
+    let data_store_config = DataStoreConfig {
+        available_blocks_cache_capacity: 1000,
+        message_id_boundary: 100_000,
+        periodic_maintenance_interval: Duration::from_millis(30),
+        request_block_after: Duration::from_millis(50),
+    };
+
+    let (mut data_store, network) = DataStore::<
+        Block,
+        TestClient,
+        Backend,
+        TestBlockRequester<Block>,
+        TestData,
+        mpsc::UnboundedReceiver<TestData>,
+    >::new(
+        client.clone(),
+        block_requester,
+        data_store_config,
+        test_network,
+    );
+    let (exit_data_store_tx, exit_data_store_rx) = oneshot::channel();
+
+    (
+        async move {
+            data_store.run(exit_data_store_rx).await;
+        },
+        client,
+        DataStoreChannels {
+            network: Box::new(network),
+            block_requests_rx,
+            network_tx,
+            exit_data_store_tx,
+        },
+    )
+}
+
+async fn import_blocks(
+    client: &mut Arc<TestClient>,
+    n: u32,
+    finalize: bool,
+) -> Vec<AlephDataFor<Block>> {
+    let mut blocks = vec![];
+    for _ in 0..n {
+        let block = client
+            .new_block(Default::default())
+            .unwrap()
+            .build()
+            .unwrap()
+            .block;
+        if finalize {
+            client
+                .import_as_final(BlockOrigin::Own, block.clone())
+                .await
+                .unwrap();
+        } else {
+            client
+                .import(BlockOrigin::Own, block.clone())
+                .await
+                .unwrap();
+        }
+        blocks.push(AlephData::new(block.header.hash(), block.header.number));
+    }
+    blocks
+}
+
+#[tokio::test]
+async fn sends_messages_with_imported_blocks() {
+    let (
+        task_handle,
+        mut client,
+        DataStoreChannels {
+            mut network,
+            block_requests_rx: _,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+
+    let blocks = import_blocks(&mut client, 4, false).await;
+
+    network_tx.unbounded_send(blocks.clone()).unwrap();
+
+    let message = network.next().await.unwrap();
+    assert_eq!(message.included_blocks(), blocks);
+
+    exit_data_store_tx.send(()).unwrap();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn sends_messages_after_import() {
+    let (
+        task_handle,
+        mut client,
+        DataStoreChannels {
+            mut network,
+            block_requests_rx: _,
+            network_tx,
+            exit_data_store_tx,
+        },
+    ) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+
+    // Build block
```

Comment:
You repeat this quite a lot, maybe abstract it away into a function.

Reply:
Done